sql: combination of DATABASE and USER statements wedges server

See #29089 for a quick test reproduction. RSG tests have been failing because combining a bunch of database and user {create, drop, alter} statements in parallel wedges a server. We had a bug similar to this just a few weeks back that was triggered just by database statements. Tagging @vivekmenezes and @jordanlewis because I'm unsure who wants to triage this.

Found with RSG.

Another reproduction of this on #29255:
In the TC output linked below, it looks like the only two running are the test runner waiting for result? There are lots of goroutines waiting on the db cache mutex though and also several on txn push. Hopefully someone more familiar with the code involved can tell from those stack traces if this looks alarming or not.

```
panic: timeout: "CREATE USER PLACEHOLDER WITH PASSWORD PLACEHOLDER".
currently executing: map[
CREATE USER PLACEHOLDER WITH PASSWORD PLACEHOLDER:1
ALTER SEQUENCE ONLY ident . MINVALUE RENAME TO BOOLEAN . VIRTUAL . DAY:1 
DROP DATABASE IF EXISTS ident RESTRICT:1 
CREATE USER 'string' PASSWORD RESTORE:1 
RESET CLUSTER SETTING EXPERIMENTAL_AUDIT . ident . HAVING . REAL . ident:1 
CREATE DATABASE IF NOT EXISTS ident:15
```

https://teamcity.cockroachdb.com/viewLog.html?buildId=867074&tab=buildResultsDiv&buildTypeId=Cockroach_Nightlies_NightlySuite

--- a/pkg/sql/database.go
+++ b/pkg/sql/database.go
@@ -253,6 +253,24 @@ func (dc *databaseCache) getDatabaseID(
 	return desc.ID, nil
 }
 
+// getCachedDatabaseID returns the ID of a database given its name
+// from the cache. This method never goes to the store to resolve
+// the name to id mapping. Returns 0 if the name to id mapping or
+// the database descriptor are not in the cache.
+func (dc *databaseCache) getCachedDatabaseID(ctx context.Context, name string) (sqlbase.ID, error) {
+	if id := dc.getID(name); id != 0 {
+		return id, nil
+	}
+
+	desc, err := dc.getCachedDatabaseDesc(name, false /*required*/)
+	if err != nil || desc == nil {
+		// desc can be nil if required == false and the database was not found.
+		return 0, err
+	}
+
+	return desc.ID, nil
+}
+
 // renameDatabase implements the DatabaseDescEditor interface.
 func (p *planner) renameDatabase(
 	ctx context.Context, oldDesc *sqlbase.DatabaseDescriptor, newName string,

--- a/pkg/sql/table.go
+++ b/pkg/sql/table.go
@@ -400,8 +400,7 @@ func (tc *TableCollection) waitForCacheToDropDatabases(ctx context.Context) {
 		tc.dbCacheSubscriber.waitForCacheState(
 			func(dc *databaseCache) bool {
 				// Resolve the database name from the database cache.
-				dbID, err := dc.getDatabaseID(ctx,
-					tc.leaseMgr.execCfg.DB.Txn, uc.name, false /*required*/)
+				dbID, err := dc.getCachedDatabaseID(ctx, uc.name)
 				if err != nil || dbID == 0 {
 					// dbID can still be 0 if required is false and
 					// the database is not found. Swallowing error here

